### ISSUE TITLE
IllegalArgumentException: discriminator or property not found for oneOf schema

### ISSUE THREAD

**Issue by jmini** [OpenAPITools/openapi-diff#29]
I am doing a test with a schema containing a oneOf schema (see documentation [Data Models (Schemas) > oneOf, anyOf, allOf, not](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/))

I think my specification is valid and makes sense (please do not hesitate to discuss it)

In this test case, one of the `oneOf` referenced schemas was removed between the old and the new version. Depending on the operation (GET or POST) the `ComposedSchema` is located: 

* in the input (request body of the POST method) 
* in the output (answer of the GET method).

My feeling is that a removal in the output is still backward compatible (GET Operation), and not in the input (POST Operation).

But for the moment the diff tool is crashing before (see stacktrace)

#### Old Spec:

```
openapi: 3.0.1
info:
  title: oneOf test
  version: '1.0'
servers:
  - url: 'http://localhost:8000/'
paths:
  /state:
    get:
      operationId: getState
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                oneOf:
                  - $ref: '#/components/schemas/A'
                  - $ref: '#/components/schemas/B'
                  - $ref: '#/components/schemas/C'
    post:
      operationId: update
      requestBody:
        content:
            application/json:
              schema:
                oneOf:
                  - $ref: '#/components/schemas/A'
                  - $ref: '#/components/schemas/B'
                  - $ref: '#/components/schemas/C'
        required: true
      responses:
        '201':
          description: OK
components:
  schemas:
    A:
      type: object
      properties:
        message:
          type: string
    B:
      type: object
      properties:
        description:
          type: string
        code:
          type: integer
          format: int32
    C:
      type: object
      properties:
        case:
          type: integer
          format: int64
```

#### New Spec:

```
openapi: 3.0.1
info:
  title: oneOf test
  version: '1.0'
servers:
  - url: 'http://localhost:8000/'
paths:
  /state:
    get:
      operationId: getState
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                oneOf:
                  - $ref: '#/components/schemas/A'
                  - $ref: '#/components/schemas/B'
    post:
      operationId: update
      requestBody:
        content:
            application/json:
              schema:
                oneOf:
                  - $ref: '#/components/schemas/A'
                  - $ref: '#/components/schemas/B'
        required: true
      responses:
        '201':
          description: OK
components:
  schemas:
    A:
      type: object
      properties:
        message:
          type: string
    B:
      type: object
      properties:
        description:
          type: string
        code:
          type: integer
          format: int32
    C:
      type: object
      properties:
        case:
          type: integer
          format: int64
```

#### Stacktrace:

```
Exception in thread "main" java.lang.IllegalArgumentException: discriminator or property not found for oneOf schema
	at com.qdesrame.openapi.diff.compare.schemadiffresult.ComposedSchemaDiffResult.diff(ComposedSchemaDiffResult.java:41)
	at com.qdesrame.openapi.diff.compare.SchemaDiff.computeDiff(SchemaDiff.java:90)
	at com.qdesrame.openapi.diff.compare.SchemaDiff.computeDiff(SchemaDiff.java:16)
	at com.qdesrame.openapi.diff.compare.ReferenceDiffCache.cachedDiff(ReferenceDiffCache.java:39)
	at com.qdesrame.openapi.diff.compare.SchemaDiff.diff(SchemaDiff.java:62)
	at com.qdesrame.openapi.diff.compare.ContentDiff.diff(ContentDiff.java:38)
	at com.qdesrame.openapi.diff.compare.ResponseDiff.computeDiff(ResponseDiff.java:38)
	at com.qdesrame.openapi.diff.compare.ResponseDiff.computeDiff(ResponseDiff.java:15)
	at com.qdesrame.openapi.diff.compare.ReferenceDiffCache.cachedDiff(ReferenceDiffCache.java:39)
	at com.qdesrame.openapi.diff.compare.ResponseDiff.diff(ResponseDiff.java:28)
	at com.qdesrame.openapi.diff.compare.ApiResponseDiff.diff(ApiResponseDiff.java:32)
	at com.qdesrame.openapi.diff.compare.OperationDiff.diff(OperationDiff.java:41)
	at com.qdesrame.openapi.diff.compare.PathDiff.diff(PathDiff.java:31)
	at com.qdesrame.openapi.diff.compare.PathsDiff.lambda$diff$2(PathsDiff.java:50)
	at java.util.LinkedHashMap$LinkedKeySet.forEach(LinkedHashMap.java:559)
	at com.qdesrame.openapi.diff.compare.PathsDiff.diff(PathsDiff.java:36)
	at com.qdesrame.openapi.diff.compare.OpenApiDiff.compare(OpenApiDiff.java:92)
	at com.qdesrame.openapi.diff.compare.OpenApiDiff.compare(OpenApiDiff.java:66)
	at com.qdesrame.openapi.diff.OpenApiCompare.fromSpecifications(OpenApiCompare.java:99)
	at com.qdesrame.openapi.diff.OpenApiCompare.fromLocations(OpenApiCompare.java:88)
	at com.qdesrame.openapi.diff.OpenApiCompare.fromFiles(OpenApiCompare.java:65)
	at com.qdesrame.openapi.diff.OpenApiCompare.fromFiles(OpenApiCompare.java:53)
```

**Comment by adarshsharma** [comment]
You are right about the schemas. They are valid specs. Currently we have implemented the code such that it expects discriminator field for oneOf type of ComposedSchemas because it would be  hard to deserialize the objects into concrete classes without the propertyName. It is somewhat similar to having Json Subtype and mentioning the field name using which you can deserialize your object.

**Comment by adarshsharma** [comment]
And yes the below expectation is correct. 
"My feeling is that a removal in the output is still backward compatible (GET Operation), and not in the input (POST Operation)."

**Comment by jmini** [comment]
Thank you for the quick feedback. I had a look at the [discriminator object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#discriminatorObject) in the specification.

The diff is working with those specs:

#### Old Spec:
```
openapi: 3.0.1
info:
  title: oneOf test
  version: '1.0'
servers:
  - url: 'http://localhost:8000/'
paths:
  /state:
    get:
      operationId: getState
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                oneOf:
                  - $ref: '#/components/schemas/A'
                  - $ref: '#/components/schemas/B'
                discriminator:
                  propertyName: realtype
                  mapping:
                    a-type: '#/components/schemas/A'
                    b-type: '#/components/schemas/B'
    post:
      operationId: update
      requestBody:
        content:
            application/json:
              schema:
                oneOf:
                  - $ref: '#/components/schemas/A'
                  - $ref: '#/components/schemas/B'
                discriminator:
                  propertyName: realtype
                  mapping:
                    a-type: '#/components/schemas/A'
                    b-type: '#/components/schemas/B'
        required: true
      responses:
        '201':
          description: OK
components:
  schemas:
    A:
      type: object
      properties:
        message:
          type: string
    B:
      type: object
      properties:
        description:
          type: string
        code:
          type: integer
          format: int32
```

#### New Spec:
```
openapi: 3.0.1
info:
  title: oneOf test
  version: '1.0'
servers:
  - url: 'http://localhost:8000/'
paths:
  /state:
    get:
      operationId: getState
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                oneOf:
                  - $ref: '#/components/schemas/A'
                  - $ref: '#/components/schemas/B'
                  - $ref: '#/components/schemas/C'
                discriminator:
                  propertyName: realtype
                  mapping:
                    a-type: '#/components/schemas/A'
                    b-type: '#/components/schemas/B'
                    c-type: '#/components/schemas/C'
    post:
      operationId: update
      requestBody:
        content:
            application/json:
              schema:
                oneOf:
                  - $ref: '#/components/schemas/A'
                  - $ref: '#/components/schemas/B'
                  - $ref: '#/components/schemas/C'
                discriminator:
                  propertyName: realtype
                  mapping:
                    a-type: '#/components/schemas/A'
                    b-type: '#/components/schemas/B'
                    c-type: '#/components/schemas/C'
        required: true
      responses:
        '201':
          description: OK
components:
  schemas:
    A:
      type: object
      properties:
        message:
          type: string
    B:
      type: object
      properties:
        description:
          type: string
        code:
          type: integer
          format: int32
    C:
      type: object
      properties:
        case:
          type: integer
          format: int64
```

#### Question

Did I understood your input correctly?

**Comment by adarshsharma** [comment]
This seems fine. Only missing part seems to be the property name 'realtype'
in each of the schemas (A, B and C). Currently there is no strict check of
this property to be available in all the oneOf type of refs.

Thanks,
Adarsh

**Comment by quen2404** [comment]
Hi guys,
i have this problem for one API contract of my company. After some reflection, i think i will implement this solution if there isn't a discriminator:
Compare each schema (item 1 vs item 1, etc...) and notify if list length change.
What do you think about this ?

**Comment by adarshsharma** [comment]
It seems fine if there isn't a discriminator. We just need to specify that
reordering of the objects will not be taken care by the diff.

**Comment by jmini** [comment]
In my case I have switched to the pattern with discriminator. I think it is good to have both solutions implemented.

**Comment by jmini** [comment]
@quen2404
Did you made any progress on this?

I have a new case where I could use a solution without discriminator. I have a field that can return an `id` (as `String`) or an expanded `Object`. I think that the notation in OpenAPI is:

```
        billingAddress:
          oneOf:
          - type: string
          - $ref: '#/components/schemas/Address'
```

This is similar to what Stripe is doing in their [spec](https://raw.githubusercontent.com/stripe/openapi/master/openapi/spec3.json):
```
      "backwards_compatible_platform_earning": {
        "properties": {
          "account": {
            "anyOf": [
              {
                "type": "string"
              },
              {
                "$ref": "#/components/schemas/account"
              }
            ],
            "description": "ID of the Stripe account this fee was taken from."
            }
          },
```
They are using `anyOf` but I think that `oneOf` is more correct.

In theses cases, there is definitively no discriminator, because the type of the object is used to determine if the object is expanded or not.

**Comment by jmini** [comment]
I will use a very quick & ugly fix for the moment.
If you have not started to work on it, I can propose something (just tell me). 

---

I think a first step is to set `changedSchema.setDiscriminatorPropertyChanged(true)` if the discriminator is set in one spec and not in the other.

Then checking if one of the `oneOf` list is empty: this is currently implemented, but I think 
that the current result of having `DiscriminatorPropertyChanged` set to `true` when both discriminators are the same but one  of the two list is empty is wrong.

Then distinguish:

* Comparison if both discriminator are not set (no logic, to be discussed [1])
* Comparison if both discriminator are set (current logic)

---

[1] I am not sure I understood what you have in mind in your previous messages. In a first iteration, we can just compare the size of each lists.

* If size is the same left and right => no change.
* If size is different => change.

---

I wait for your feedback.

**Comment by quen2404** [comment]
Hello, no i didn't start to work on this problem, so you can do it :) 
Yes your logic is good.
About [1]:
Yes you can compare the list size, but i want to compare elements in the list too. 
oldSchema.oneOf[0] against newSchema.oneOf[0] etc...

**Comment by NikitaMarushkai** [comment]
Hello guys! Have this been fixed already? Coz in version 1.2.0 the issue with missing discriminator is still exists

**Comment by quen2404** [comment]
I think this is still an issue in version `2.0.0-SNAPSHOT`. I should work on it.

**Comment by fanchen1988** [comment]
> I think this is still an issue in version `2.0.0-SNAPSHOT`. I should work on it.

Hi @quen2404 , is this fixed? or when will it be fixed? I still can reproduce this with the docker image

**Comment by joschi** [comment]
Fixed in 726740a8c2d95d6aaaafbd9efe4cbc7a54f72b46.
